python/cudf_polars/cudf_polars/dsl/ir.py

```diff
@@ -71,13 +104,18 @@ def broadcast(
             return list(columns)
         nrows = target_length
     elif len(lengths) == 1:
-        if target_length is not None:
-            assert target_length in lengths
+        if target_length is not None and target_length not in lengths:
+            raise RuntimeError(
+                "Cannot broadcast columns of length "
+                f"{lengths.pop()} to {target_length=}"
+            )
         return list(columns)
     else:
         (nrows,) = lengths - {1}
-        if target_length is not None:
-            assert target_length == nrows
+        if target_length is not None and target_length != nrows:
+            raise RuntimeError(
+                f"Cannot broadcast columns of length {nrows} to {target_length=}"
+            )
```

Comment:
I _think_ some of this logic can be simplified by combining some of the `target_length != None` cases. What about something like

```python
nrows = target_length
if nrows is None:
    if lengths == {1}:
        # we have one scalar, and we're not broadcasting it 
        return list(columns)
    # target_length was never provided, so n_rows must be the other unique length 
    nrows = (lengths - {1}).pop()


elif len(lengths - {1}) == 1 and target_length not in lengths:
    # target_length must have been passed to reach this elif
    # only continue if passed length is the single common length amongst passed columns
    raise RuntimeError("column mismatch....")
        
```
This way we just say "if we have a 1 and a length, the passed `target_length` must be one of those two numbers"

Reply:
I'll think, but you are right, I think this can be cleaned up a bit.

Reply:
Done.